Refactor Enocean part 1

jduquennoy opened the pull request:
Hi guys,

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR is the first of a serie, aiming at a refactoring / modernization of the EnOcean integration.
This PR mostly reorganized some code, splitting it is different files to increase readability, and adds a config flow to configure the EnOcean dongle.
The config flow will try to auto-detect USB dongles, based on a fixed list of globs, or will request a manual user entry if no candidate path is found.

The old text-based configuration is still operational with those changes.

Next steps of the refatoring will include
- auto-creation of entities when receiving a teach-in packet or a learn packet
- evolutions of the entities to rely more on the external library to decode and create packets
- support of new devices (a window sensor notably), thanks to the previous point
- automated tests

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: 
- This PR is related to issue: 
- This PR is related to PR: #33094
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13853

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally. (at least with the devices I own)
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass** (Tox is failing on my config, working on that)
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works. (the enOcean integration does not have tests as of now. Adding some will be part of next steps in the refactoring).

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [X] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jduquennoy committed:
First step of an EnOcean integration refactoring, including code reorganisation and support of a setup config flow

homeassistant commented:
Hi @jduquennoy, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#35927).

Once you do that we will be able to review and accept this pull request.

Thanks!

jduquennoy committed:
Moved title to root of strings file

jduquennoy committed:
Fixed pre-commit checks failures

jduquennoy committed:
Fixed linter errors

jduquennoy commented:
Hum, strange: the linter executed by the CI is complaining that some formatted strings are not valid:
`Use % formatting in logging functions and pass the % parameters as arguments (logging-format-interpolation)`
Originally, those strings were written as `"string {}".format(arg)`, which was the recommanded way if I trust what I read about it.
With the recent changes, this syntax is changed by the pre-commit hook to `f"string {arg}"`, and this does not pass the CI checks.

What have I missed here, what syntax should I use to have both systems agree ?

MartinHjelmare commented:
1. %-formatting in logging messages.
2. str.format in constants.
3. f-strings in all other cases.

jduquennoy committed:
Updated formatted string format in logs

jduquennoy committed:
Removed leftover comment

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+        if new_state == DongleSetupStates.VALIDATE:
+            LOGGER.debug("Config step: validate")
+            dongle_path = user_input[CONF_DEVICE]
+            if path.exists(dongle_path) and not path.isdir(dongle_path):
```
I/O. Please use the executor. Probably break out a separate validation function that we can schedule on the executor.

Before creating the entry we also want to check that we can connect to the device.

jduquennoy replied:
Done.
I added a method in EnOceanDongle to validate the path, that I call using async_add_executor_job.

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+
+        if new_state == DongleSetupStates.CREATE_ENTRY:
+            LOGGER.debug("Config step: create entry")
+            await self.async_set_unique_id(user_input[CONF_DEVICE])
```
Is the path really unique and stable? If the USB is connected a second time the device path might change, but the device is still the same. I don't think we can use the path and our unique_id api together.

We only allow a single config entry and there's no discovery step so I don't think we need a unique_id for the config entry.

jduquennoy replied:
Ok, removed the call to set_unique_id

MartinHjelmare commented on homeassistant/components/enocean/device.py:
```diff
@@ -0,0 +1,41 @@
+from homeassistant.helpers.entity import Entity
+
+
+class EnOceanDevice(Entity):
```
Please name it `SomethingEntity`. Ie end with `Entity`.

jduquennoy replied:
Ok, renamed to EnOceanEntity

MartinHjelmare commented on homeassistant/components/enocean/device.py:
```diff
@@ -0,0 +1,41 @@
+
+
+class EnOceanDevice(Entity):
+    """Parent class for all devices associated with the EnOcean component."""
```
It's a parent class for all entities not devices.

jduquennoy replied:
Comment updated

MartinHjelmare commented on homeassistant/components/enocean/dongle.py:
```diff
@@ -0,0 +1,66 @@
+    """
+
+    @classmethod
+    def detect(cls):
```
We're not using the `cls`. Maybe make it a static method instead?

jduquennoy replied:
Indeed. I changed it to a static method.

MartinHjelmare commented on homeassistant/components/enocean/dongle.py:
```diff
@@ -0,0 +1,66 @@
+    def __init__(self, hass, serial_path):
+        """Initialize the EnOcean dongle."""
+
+        LOGGER.debug("Creating dongle for path %s", serial_path)
```
We don't want side effects like logging in init method. Please remove or move the logging.

jduquennoy replied:
Log removed.

MartinHjelmare commented on homeassistant/components/enocean/dongle.py:
```diff
@@ -0,0 +1,66 @@
+        """Initialize the EnOcean dongle."""
+
+        LOGGER.debug("Creating dongle for path %s", serial_path)
+        self.__communicator = SerialCommunicator(
```
Please use a single underscore for private instance attributes. Double underscore activates name mangling which we don't need here.

jduquennoy replied:
Renamed with only one underscore.

MartinHjelmare commented on homeassistant/components/enocean/dongle.py:
```diff
@@ -0,0 +1,66 @@
+        self.serial_path = serial_path
+        self.identifier = basename(normpath(serial_path))
+        self.hass = hass
+        async_dispatcher_connect(hass, SIGNAL_SEND_MESSAGE, self._send_message_callback)
```
I'd probably add a dedicated setup method where we can connect the dispatcher and start the connector.

jduquennoy replied:
I have added an async setup method, that start the serial communicator and connects to the send_message handler.

MartinHjelmare commented on homeassistant/components/enocean/dongle.py:
```diff
@@ -0,0 +1,66 @@
+        self.__communicator = SerialCommunicator(
+            port=serial_path, callback=self.callback
+        )
+        self.__communicator.start()
```
I assume starting the communicator is a side effect.

jduquennoy replied:
In the init, I guess it is yes. It has been moved to an async_setup method.

MartinHjelmare commented on homeassistant/components/enocean/manifest.json:
```diff
@@ -2,6 +2,12 @@
+  "requirements": [
+    "enocean==0.50"
+  ],
+  "dependencies": [],
```
Please remove empty items.

jduquennoy replied:
Removed

MartinHjelmare commented on homeassistant/helpers/translation.py:
```diff
@@ -21,7 +21,7 @@
 
 MOVED_TRANSLATIONS_DIRECTORY_MSG = (
-    "%s: the '.translations' directory has been moved, the new name is 'translations', "
+    "%s: the 'translations' directory has been moved, the new name is 'translations', "
```
Please don't change this file.

jduquennoy replied:
Ouch, sorry ! I trusted the refactoring of pyCharm too much, sorry for that !
I reverted the unwanted changes.

MartinHjelmare commented on script/hassfest/translations.py:
```diff
@@ -31,15 +31,15 @@
 def check_translations_directory_name(integration: Integration) -> None:
     """Check that the correct name is used for the translations directory."""
-    legacy_translations = integration.path / ".translations"
+    legacy_translations = integration.path / "translations"
```
Please don't change this file.

jduquennoy replied:
reverted

MartinHjelmare reviewed (changes requested):
The config flow requires 100% test coverage. I recommend using our scaffold script for config flow to get started with tests and also see our best practice patterns for the config flow.

jduquennoy committed:
Multiple changes after PR change requests.
Using an import flow for yaml config, removed unnecessary logs, added proper unload in __init__ and EnOceanDongle
Replaced config state machine by several flows.
Serial port validity check done in the EnOceanDongle class asynchronously, removed unique ID from config flow
Multiple cosmetic changes

jduquennoy commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,60 @@
-
-DOMAIN = "enocean"
-DATA_ENOCEAN = "enocean"
+from ... import config_entries, core
```
All imports fixed

jduquennoy commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,60 @@
-
-class EnOceanDongle:
-    """Representation of an EnOcean dongle."""
+    LOGGER.debug("setup with config")
```
Unnecessary logs removed

jduquennoy commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,60 @@
-        """Initialize the EnOcean dongle."""
+    # support for text-based configuration (legacy)
+    if DOMAIN in config:
+        serial_dev = config[DOMAIN].get(CONF_DEVICE)
```
Change done

jduquennoy commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,60 @@
-from enocean.utils import combine_hex
 import voluptuous as vol
 
+from homeassistant.components.enocean.const import DATA_ENOCEAN, DOMAIN, LOGGER
```
Fixed

jduquennoy commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,60 @@
+    # support for text-based configuration (legacy)
+    if DOMAIN in config:
+        serial_dev = config[DOMAIN].get(CONF_DEVICE)
+        enocean_dongle = EnOceanDongle(hass, serial_dev)
```
Done.
I added a check to only trigger the import flow if there is no config entry.

jduquennoy commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,60 @@
+    This will read a flow-based configuration, but only if
+    a configuration was not found in the text-based configuration.
+    """
+    LOGGER.debug("async setup")
```
Removed

jduquennoy commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,60 @@
 
-    def _message_received_callback(self, packet):
-        """Handle incoming packets."""
+    if CONF_DEVICE not in config_entry.data:
```
Removed

MartinHjelmare commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,66 @@
-            )
-        )
+    This will read a flow-based configuration, but only if
+    a configuration was not found in the text-based configuration.
```
This isn't correct. `async_setup_entry` will be called for all config entries created.

jduquennoy replied:
Comment updated.

MartinHjelmare commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,66 @@
+    This will read a flow-based configuration, but only if
+    a configuration was not found in the text-based configuration.
+    """
+    if DATA_ENOCEAN in hass.data and hass.data[DATA_ENOCEAN] is EnOceanDongle:
+        LOGGER.warning("Dongle already configured, skipping configuration")
+        return True
 
```
```suggestion
```

jduquennoy replied:
Suggestion applied

jduquennoy commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,60 @@
-        self.hass.helpers.dispatcher.dispatcher_send(SIGNAL_SEND_MESSAGE, packet)
+async def async_unload_entry(hass, config_entry):
+    """Unload ENOcean config entry."""
+    LOGGER.debug("Unloading %s", DOMAIN)
```
Removed

MartinHjelmare commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,66 @@
 
-    def _message_received_callback(self, packet):
-        """Handle incoming packets."""
+    if DATA_ENOCEAN not in hass.data:
+        hass.data[DATA_ENOCEAN] = {}
```
```suggestion
    enocean_data = hass.data.setdefault(DATA_ENOCEAN, {})
```

jduquennoy replied:
In ended up having to create a variable to store the default value, as setdefault was returning None when receiving the emty dict literal directly.
This was only happening when creating the integration though, not when starting hass with an already configured integration.
I am still investigating to understand why.

jduquennoy replied:
Suggestion applied

MartinHjelmare commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,66 @@
-        if packet.sender_int == combine_hex(self.dev_id):
-            self.value_changed(packet)
+    usb_dongle = EnOceanDongle(hass, config_entry.data[CONF_DEVICE])
+    hass.data[DATA_ENOCEAN] = usb_dongle
```
```suggestion
    enocean_data[ENOCEAN_DONGLE] = usb_dongle
```

jduquennoy replied:
Sorry for the pretty long delay since my last activity on this PR.
I have applied the different changes, but I end up with a strange behavior: enocean_data is supposed to have a value after the call to setdefault. But when assigning usb_dongle to it two lines (including an await) later, it is null.
There is certainly a python behavior I have misunderstood. I am investigating, I will not resolve this conversation until I understand what is happening.

jduquennoy replied:
fixed after updating my dev env to python 3.8.

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+
+    async def async_step_import(self, data=None):
+        """Import a yaml configuration."""
+        return self.async_create_entry(title="EnOcean", data=data)
```
We want to check the connection to the dongle before creating the entry for import too.

jduquennoy replied:
Added the check. If it fails, I return async_abort. It does not seem to display an error in the UI. I also have added a warning log.

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+            return self.async_abort(reason="single_instance_allowed")
+
+        if user_input is None:
+            return await self.async_step_detect()
+        elif user_input[CONF_DEVICE] == self.MANUAL_PATH_VALUE:
+            return await self.async_step_manual(user_input)
+        else:
+            return await self.async_step_validate(user_input)
```
```suggestion
            return await self.async_step_detect()
        if user_input[CONF_DEVICE] == self.MANUAL_PATH_VALUE:
            return await self.async_step_manual(user_input)
        return await self.async_step_validate(user_input)
```

jduquennoy replied:
Suggestion applied

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+            return await self.async_step_manual(user_input)
+
+        bridges.append(self.MANUAL_PATH_VALUE)
+        return self.async_show_form(
+            step_id="validate",
+            data_schema=vol.Schema({vol.Required(CONF_DEVICE): vol.In(bridges)}),
+        )
```
Each form belongs to a single step. We change step by returning and awaiting the next step method. We don't go directly to another step's form.

jduquennoy replied:
Ok, both forms are updated, and the logic of the steps are updated accordingly.

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+            data_schema=vol.Schema({vol.Required(CONF_DEVICE): vol.In(bridges)}),
+        )
+
+    async def async_step_manual(self, user_input=None, errors=None):
```
Errors should not be passed between steps. The errors are specific to the form of the step that received the errors.
```suggestion
    async def async_step_manual(self, user_input=None):
```

jduquennoy replied:
errors parameter is removed.

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+
+    async def async_step_manual(self, user_input=None, errors=None):
+        """Request manual USB dongle path."""
+        default_value = None
```
```suggestion
        errors = {}
        default_value = None
```

jduquennoy replied:
Suggestion applied

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+        default_value = None
+        if user_input is not None and user_input[CONF_DEVICE] != self.MANUAL_PATH_VALUE:
+            default_value = user_input[CONF_DEVICE]
+        return self.async_show_form(
+            step_id="validate",
+            data_schema=vol.Schema(
+                {vol.Required(CONF_DEVICE, default=default_value): str}
+            ),
+            errors=errors,
+        )
```
See above. A step should not use another steps' form.

jduquennoy replied:
Form updated

jduquennoy commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+
+    VERSION = 1
+    MANUAL_PATH_VALUE = "Custom path"
+    CONNECTION_CLASS = CONNECTION_ENOCEAN
```
Changed to for CONN_CLASS_ASSUMED.

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,94 @@
+
+        return await self.async_step_create_entry(user_input)
+
+    async def async_step_create_entry(self, user_input):
```
This doesn't have to be a step method since there's no form. It can be a helper method.

jduquennoy replied:
Changed to a non-step method

jduquennoy commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+        self.dongle_path = None
+        self.discovery_info = None
+
+    async def async_step_init(self, user_input=None):
```
Removed.

jduquennoy commented on homeassistant/components/enocean/strings.json:
```diff
@@ -0,0 +1,20 @@
+      "invalid_dongle_path": "No dongle found for this path."
+    },
+    "abort": {
+      "dongle_already_configured": "An enocean dongle is already configured. If it is not visible in the UI, check the configuration.yaml file."
```
dongle_already_configured replaced by single_instance_allowed

jduquennoy commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+        It uses a state machine to walk through the different steps of the
+        configuration.
+        """
+        if self._async_current_entries() or DOMAIN in self.hass.data:
```
Removed

MartinHjelmare commented on homeassistant/components/enocean/dongle.py:
```diff
@@ -0,0 +1,85 @@
+        self.hass = hass
+        self.dispatcher_disconnect_handle = None
+
+    def __del__(self):
```
Let's make a normal method that cleans up, to make it more explicit.

jduquennoy replied:
ok, I converted the __del__ to an unload method, that is called explicitely from async_unload_entry

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -3,12 +3,12 @@
     PLATFORM_SCHEMA,
     BinarySensorEntity,
 )
+from homeassistant.components.enocean.device import EnOceanEntity
```
Intra package imports should be relative.

jduquennoy replied:
ouch, again. Corrected, and IDE configuration updated to avoid reproducing this error.

jduquennoy commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+
+        if new_state == DongleSetupStates.SELECT_AUTO:
+            LOGGER.debug("Config step: select auto")
+            bridges = EnOceanDongle.detect()
```
Done, thanks for the hint !

jduquennoy commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+
+        return await self.set_dongle_setup_state(config_state, user_input)
+
+    async def set_dongle_setup_state(self, new_state, user_input=None, errors=None):
```
Done. The states are now steps.

MartinHjelmare commented on homeassistant/components/enocean/dongle.py:
```diff
@@ -0,0 +1,85 @@
+            test_port = serial.Serial(path, 57600, timeout=0.1)
+            test_port.close()
+            return True
+        except Exception as e:
```
Please catch specific errors. I think we can limit it to `SerialException`.

https://pyserial.readthedocs.io/en/latest/pyserial_api.html#exceptions

jduquennoy replied:
Done

jduquennoy commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,109 @@
+            default_value = None
+            if (
+                user_input is not None
+                and CONF_DEVICE in user_input
```
Ok, check removed.

MartinHjelmare commented on homeassistant/components/enocean/dongle.py:
```diff
@@ -0,0 +1,85 @@
+    def validate_path(path: str):
+        """Return True if the provided path points to a valid serial port, False otherwise."""
+        try:
+            test_port = serial.Serial(path, 57600, timeout=0.1)
```
Could we make a `SerialCommunicator` instance instead and try the same concept?

jduquennoy replied:
SerialCommunicator is a subclass of Thread, I was thinking that starting a thread for that is a bit overkill.
But indeed, I can use SerialCommunicator.

jduquennoy replied:
Change done

MartinHjelmare commented on homeassistant/components/enocean/strings.json:
```diff
@@ -0,0 +1,19 @@
+      "invalid_dongle_path": "No dongle found for this path."
+    },
+    "abort": {
+    }
```
```suggestion
      "single_instance_allowed": "[%key:common::config_flow::abort::single_instance_allowed%]"
    }
```

jduquennoy replied:
Suggestion applied.

MartinHjelmare commented on homeassistant/components/enocean/strings.json:
```diff
@@ -0,0 +1,19 @@
+  "config": {
+    "flow_title": "ENOcean setup",
+    "step": {
+      "init": {
```
Add all the steps with forms and the specific form data names.

jduquennoy replied:
strings now contains two steps: detect and manual

jduquennoy committed:
Multiple changes after PR change requests

jduquennoy committed:
Added variable to store default value, as setdefault was caught returning None when the empty dict literal was passed as an argument

MartinHjelmare commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,59 @@
+):
+    """Set up an EnOcean dongle for the given entry."""
+    default_enocean_data = {}
+    enocean_data = hass.data.setdefault(DATA_ENOCEAN, default_enocean_data)
```
```
Python 3.8.2 (default, Feb 26 2020, 02:56:10) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.13.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: data = {}                                                                                                                                                               

In [2]: test = data.setdefault("special_data", {})                                                                                                                              

In [3]: test                                                                                                                                                                    
Out[3]: {}

In [4]: data                                                                                                                                                                    
Out[4]: {'special_data': {}}
```

jduquennoy replied:
Yes, I did the same test here, with the same result (but with a python 3.7.7 interpreter).
As I mentionned, it also only happens when creating a new integration, not when setting up an already configured integration.
I am currently upgrading to python 3.8.2, hoping it will help.

jduquennoy replied:
It does work perfectly with python 3.8.2.
Use of the literal commited.

jduquennoy committed:
Literal used directly

jduquennoy committed:
Added tests for EnOcean config flows, changed static methods to bundle methods for bundle

jduquennoy commented:
With the latest commit, code coverage is 100% for the config flows.

jduquennoy committed:
Updated variable name

jduquennoy committed:
Added missing mock to test, replaced repeated magic strings by constants

jduquennoy committed:
Changed imports to avoid an unused import warning from pylint on DOMAIN

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -7,8 +7,7 @@
 
-from . import dongle
-from .const import DOMAIN, ERROR_INVALID_DONGLE_PATH, LOGGER
+from . import const, dongle
```
We prefer to import the individual names instead of whole modules.

jduquennoy replied:
Sadely, with the previous version, pylint was issuing an error "
`W0611: Unused DOMAIN imported from const`
But as you can see, DOMAIN is used on line 22. I did this change as a workaround.
I will try to find another workaround, that is not incompatible with the coding standards.

MartinHjelmare replied:
That's a bug in pylint. We don't work around it, we just disable the warning locally.

MartinHjelmare replied:
If you use our scaffold template, you can see we have this pattern already.

https://github.com/home-assistant/core/blob/08f2714e57caa017451cb4d555bfc29b8bb088f7/script/scaffold/templates/config_flow/integration/config_flow.py#L8

jduquennoy replied:
Ok, exception added.
I would not have dared to add a pylint exception on my own :-D.

jduquennoy replied:
I'm working on a commit to also import individualy detect and validate_path from dongle, but for the moment, some tests are failing after the change. Manual tests shows that the flows are still working as expected though. Investigating.

jduquennoy committed:
Adding pylint exception for unused import

MartinHjelmare commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,58 @@
 
-    def value_changed(self, packet):
-        """Update the internal state of the device when a packet arrives."""
+    return True
```
We want to forward the entry to all platforms here. I recommend using our scaffold script to see what our required and recommended patterns are.

jduquennoy replied:
Forwarding of setup and unload forwarded to platforms, following the scaffold sample.

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,108 @@
+from .const import ERROR_INVALID_DONGLE_PATH, LOGGER
+
+
+class DongleSetupStates(Enum):
```
This isn't used now.

jduquennoy replied:
Indeed, enum removed.

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,108 @@
+        """Import a yaml configuration."""
+
+        dongle_path = data[CONF_DEVICE]
+        path_is_valid = await self.hass.async_add_executor_job(
```
We can use `self.validate_enocean_conf` instead.

jduquennoy replied:
Change done

MartinHjelmare commented on homeassistant/components/enocean/config_flow.py:
```diff
@@ -0,0 +1,108 @@
+        )
+        if not path_is_valid:
+            LOGGER.warning(
+                "Cannot import yaml configuration: %s is not a valid dongle path.",
```
We don't end logging messages with period.

jduquennoy replied:
ok, trailing period removed.

MartinHjelmare commented on homeassistant/components/enocean/dongle.py:
```diff
@@ -0,0 +1,86 @@
+    def unload(self):
+        """Disconnect callbacks established at init time."""
+        if self.dispatcher_disconnect_handle:
+            self.dispatcher_disconnect_handle()
```
Set `self.dispatcher_disconnect_handle` to `None` after calling it.

jduquennoy replied:
Done

jduquennoy committed:
Added proper propagation of setup and unload to platforms, removed dead code, some syntax changes

jduquennoy commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,58 @@
+    configured_dongle = {
+        entry.data[CONF_DEVICE] for entry in hass.config_entries.async_entries(DOMAIN)
+    }
+    if configured_dongle:
```
Suggestion applied

jduquennoy commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,58 @@
+    """Unload ENOcean config entry."""
+    enocean_dongle = hass.data[DATA_ENOCEAN][ENOCEAN_DONGLE]
+    enocean_dongle.unload()
+    hass.data[DATA_ENOCEAN] = None
```
Suggestion applied

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -36,7 +37,7 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
     add_entities([EnOceanBinarySensor(dev_id, dev_name, device_class)])
```
Each platform needs to implement `async_setup_entry` to add its entities. The old `setup_platform` should not be used for this anymore.

jduquennoy replied:
I can definitely start working on the platforms also in this PR.
How do I access the yaml configuration from async_setup_entry ?
I have looked at other components configured this way (rfxtrx, tcp, template, sql), but no luck, they all rely on setup_platform.

MartinHjelmare replied:
We can't access the platform config yaml. These will go away. We'll have access to the config entry and its data.

The integration can only import config yaml to a config entry as we already do in `async_setup`.

jduquennoy replied:
I just pushed a first proposal for the import of the entities found in the YAML configuration. Does that match the expected way to do imports ?

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -26,31 +36,90 @@
+
+async def async_setup_platform(hass, config, add_entities, discovery_info=None):
+    """Convert the config entry to an entity in the registry."""
+    if not hass.config_entries.async_entries(domain=DOMAIN):
```
Here let's just store the dev_id in hass.data and then return. Do all the migration work in `async_setup_entry` instead.

jduquennoy replied:
I am now storing the config entry instead of a built entity.

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -26,31 +36,90 @@
+        YAML_BINARY_SENSORS_TO_IMPORT, []
+    )
+    for entity in configured_entities:
+        if entity.unique_id == sensor.unique_id:
```
We don't need to create an entity to get the unique_id. Just use `enocean_id_to_string` directly.

jduquennoy replied:
Done (and moved to async_setup_entry for another comment).

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -26,31 +36,90 @@
+            )
+            return
+
+    entities_to_migrate.append(EnOceanBinarySensor(dev_id, dev_name, device_class,))
```
The entities should be created in `async_setup_entry`.

jduquennoy replied:
Change done.

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -26,31 +36,90 @@
+    # Import entities from yaml configuration
+    if YAML_BINARY_SENSORS_TO_IMPORT in hass.data[DATA_ENOCEAN]:
+        entities_to_migrate = hass.data[DATA_ENOCEAN][YAML_BINARY_SENSORS_TO_IMPORT]
+        for entity in entities_to_migrate:
```
Here we should get the stored dev_ids from hass.data and generate unique_ids. Then compare those with the existing entries unique_ids from the entity registry for this config entry, and add entities for the missing dev_ids.

jduquennoy replied:
I have moved the whole logic of selecting yaml entries to migrate in async_setup_entry. The logic of checking agains the registry is now done here.

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -26,31 +36,90 @@
+
+    # Load entities from registry
+    entities_registry = await entity_registry.async_get_registry(hass)
+    entities_from_config = async_entries_for_config_entry(
```
These are entries not entities. Please rename the variable.

jduquennoy replied:
Variable renamed.

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -26,31 +36,90 @@
+    entities_from_config = async_entries_for_config_entry(
+        entities_registry, config_entry.entry_id
+    )
+    for current_entity in entities_from_config:
```
```py
for entry in entries_from_config:
```

jduquennoy replied:
Change done.

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -26,31 +36,90 @@
+    @property
+    def unique_id(self) -> Optional[str]:
+        """Return a unique ID for the device."""
+        return "{}-binary_sensor".format(enocean_id_to_string(self.dev_id))
```
We don't need to add `binary_sensor` to the unique_id. The entity registry is aware of both the integration domain and the platform domain.

jduquennoy replied:
That is cool, it does make things easier :-).
Change done.

MartinHjelmare commented on homeassistant/components/enocean/utilities.py:
```diff
@@ -0,0 +1,15 @@
+    """Return a numeric enocean id decoded from a string."""
+    numeric_id = [int(element, 16) for element in string.split(":")]
+    if len(numeric_id) != 4:
+        raise Exception(f"Invalid enocean id string: {string}")
```
Please create a new exception that inherits from `HomeAssistantError`.

jduquennoy replied:
Custom exception created and used here.

MartinHjelmare reviewed (commented):
We should also add a deprecation notice in the platform schemas using `cv.deprecated`.

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -1,56 +1,123 @@
+    )
+    for entry in entries_from_config:
+        entities_to_add.append(
+            EnOceanBinarySensor(
```
How will new entities be added for new users?

jduquennoy replied:
With this PR, the YAML config is still the only way to declare entities.
My initial objective was to do that work in the scope of another PR, to limit the size of this one (mostly focusing on migrating the integration to flows).

The EnOcean protocol processing logic will require a major change in the next steps, notably to be able to associate devices automatically. Some steps for those changes would be:
- rename the current entities to legacy, and develop new entities, that relies on the device EEP to decode and create EnOcean packets
- update the code of the dongle to get the dongle device ID at startup
- adding code to handle teach-in packets, that will allow the user to add entities by using the pairing procedure of their devices if available
- adding code to handle data packets with learn bit set, and ask the EEP of the device to the user if needed

I have prepared most of the logic for those steps. Do you think I should add it to this PR, to have a fully functional integration in the end ?

MartinHjelmare replied:
It's a bit weird that we deprecate the platform config yaml when it's needed to add new devices. Maybe the order of these PRs should be different?

A good thing is that we cut the beta last Wednesday, so there's 2.5 weeks until next beta. Will that be enough?

I don't think we should add more code to this PR.

bdurrer replied:
Note that manual configuration will always be a thing for EnOcean. Quite a lot of the most common devices do not support pairing and and auto-detection based on the payload is often not possible

MartinHjelmare replied:
Then I'm not sure we can merge this until there's a way to add new devices manually.

bdurrer replied:
Now that I think of, automatically detecting the _specific_ device type without manual intervention is not possible, because the device groups are too big / diverse.
If you want to go big, you certainly can detect new devices and then somehow have the user select the specific device type in the UI. That certainly would be cool (and a lot of work). I just wouldn't deprecate the manual configuration option before that is in place.

MartinHjelmare replied:
Ok, I think we can make this work if we use the same approach as the mqtt integration does. Have a config entry for the connection, and then config yaml for each platform.

Let's remove the migration for now.

jduquennoy replied:
In a prototype I have built to learn before proposing a PR, I was using a config flow to ask the user for a the exact type of the device. The flow was aborted at the end, after a message was sent to request the creation of the entity.
It was definitely hacky, but it was working.

I will stash the migration code tomorrow.

jduquennoy commented:
To migrate platforms such as sensor from yaml to the registry, I will need to store properties such as _range_from_, or _range_to_.
Is _capability_attributes_ the correct property to use for such additional entities data ?

MartinHjelmare commented:
No. Maybe `device_state_attributes` depending on what they mean.

UnciaPrima commented:
Hi there!

I am not a developer and I cannot help in any way, but I wanted to share my enthusiasm for this PR.

Do you plan to add support for Enocean roller shutter modules for exemple? I really wish I could use them with Home Assistant and it seems to me that there are regular requests for these modules.

Anyway, thanks a lot!

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -1,56 +1,123 @@
-        vol.Optional(CONF_DEVICE_CLASS): DEVICE_CLASSES_SCHEMA,
-    }
+PLATFORM_SCHEMA = vol.All(
+    cv.deprecated(CONF_ID),
```
Maybe mark `CONF_PLATFORM` as deprecated?

MartinHjelmare replied:
Let's keep platform yaml set up for now.

jduquennoy replied:
Migration code reverted.

MartinHjelmare commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,74 @@
-            )
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
```
I think we can remove the forwarding since we will require config yaml setup for the platforms.

If we need the config_entry.data in the platforms we can store it in hass.data.

jduquennoy replied:
Forwarding removed

MartinHjelmare commented on homeassistant/components/enocean/binary_sensor.py:
```diff
@@ -1,56 +1,123 @@
-    dev_id = config.get(CONF_ID)
-    dev_name = config.get(CONF_NAME)
-    device_class = config.get(CONF_DEVICE_CLASS)
+async def async_setup_entry(hass: HomeAssistant, config_entry, async_add_entities):
```
If there's no discovery of new devices and only manual yaml setup, we can remove `async_setup_entry` and only keep `setup_platform`.

jduquennoy replied:
async_setup_entry method removed.

Oliric commented:
> Hi there!
> 
> I am not a developer and I cannot help in any way, but I wanted to share my enthusiasm for this PR.
> 
> Do you plan to add support for Enocean roller shutter modules for exemple? I really wish I could use them with Home Assistant and it seems to me that there are regular requests for these modules.
> 
> Anyway, thanks a lot!

Hi,
In the same case: unable to help, but very thankful for this huge work!

jduquennoy commented:
> Hi there!
> I am not a developer and I cannot help in any way, but I wanted to share my enthusiasm for this PR.
> Do you plan to add support for Enocean roller shutter modules for exemple? I really wish I could use them with Home Assistant and it seems to me that there are regular requests for these modules.

Hi guys,
Thanks for the feedback, pretty motivating :-) !
The plan I have for this refactoring is to support devices that are supported by the ENOcean library. As far as I can see, the roller shutter device the the EEP A5 11 03, which does not seem to be supported as of now. But once the refactoring is done here, I will probably try to contribute to the library to add some more.

jduquennoy committed:
Removed setup_entry forwarding as the entities can only be configured using yaml

pmorange commented:
Hi,
Great to see some interest in EnOcean for HomeAssistant !

I tried to add support for more devices some time ago (https://github.com/home-assistant/core/pull/32207), but my changes were rejected because of bad implementation of the module to begin with.

Not sure if I could help with the refactoring.
Keep up the good work, and thanks from all of us using EnOcean modules.

MartinHjelmare commented on homeassistant/components/enocean/__init__.py:
```diff
@@ -1,93 +1,67 @@
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, component)
```
I'm not sure if the forward unload will work since we don't forward the entry setup. Have you tried this, also without configuring any platforms?

jduquennoy replied:
Indeed, I forgot to remove the forwarding of unload.
It is now done.

jduquennoy committed:
Removed forwarding of unload

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare reviewed (changes requested):
Sorry, I forgot. We need to include the config flow in coverage calculation. Please exclude all the other modules explicitly in `.coveragerc`.

https://github.com/home-assistant/core/blob/f69fc79fd1ba08b4cd6b557c7cf0580dfad2482c/.coveragerc#L218

jduquennoy committed:
Enabled code coverage for config flow only

MartinHjelmare commented on .coveragerc:
```diff
@@ -206,7 +206,15 @@ omit =
-    homeassistant/components/enocean/*
+    homeassistant/components/enocean/__init__.py
+    homeassistant/components/enocean/binary_sensor.py
+    homeassistant/components/enocean/
```
We can remove this line.
```suggestion
```

jduquennoy commented:
I'm surprised that two tests are failing on component cert_expiry after the last commit.
Is there a way to ask for a rerun of the tests, other than adding an empty commit ?

MartinHjelmare commented:
You can rebase on latest dev branch to let the build run again.

frenck commented:
⚠️ I was unable to find a documentation PR (matching this code change) in our [documentation repository](https://github.com/home-assistant/home-assistant.io).

Please, update the documentation and open a PR for it (be sure to create a documentation PR against the `next` branch) and link 🔗 this PR and the new documentation PR by mentioning the link to the PR's in both openings posts/descriptions.

🏷 I am adding the `docs-missing` label until this has been resolved.

jduquennoy commented:
The PR for the documentation has been added here: https://github.com/home-assistant/home-assistant.io/pull/13853
Sorry for the long delay !

MartinHjelmare committed:
Clean up coveragerc

Sebcul commented:
Nice. The Enoceans component could really need some love.
I'd love to see support for some switches/actuators (like Eltako FSR61). 
I'm currently using a custom component for this actuator (not written by me, but i can share the code if you'd like). I'd like to see that get integrated in the official component.

pmorange commented:
Hi,
Any plan on keeping the good work and working on following refactor steps ?
The first was was a necessary step it seems but it didn't add any more compatibility (at least in my case) to existing modules.

I ask here because I asked in the HA forums and nobody answered, and I can't find a way to contact @jduquennoy directly...

Cheers.

jduquennoy commented:
Sorry for the lack of answer.
There i definitely plan to continue the work, and I already have some chose to add support of a first device that supports teach-in.
Sadly, I’m in the middle of a multi-day rush, both professionally and personally. So I am not extremely active on that project those days.

I hope i will be able to focus on that again in the weeks to come.

Jérome

> Le 11 sept. 2020 à 10:18, pmorange <notifications@github.com> a écrit :
> 
> ﻿
> Hi,
> Any plan on keeping the good work and working on following refactor steps ?
> The first was was a necessary step it seems but it didn't add any more compatibility (at least in my case) to existing modules.
> 
> I ask here because I asked in the HA forums and nobody answered, and I can't find a way to contact @jduquennoy directly...
> 
> Cheers.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

bakkerv commented:
Nice to see this PR! Currently, I'm running my own branch (of an older version of HA) with support for EEP A5 11 03. When trying to merge the latest dev into my branch, I saw I had conflicts and a fast-forward was impossible.
My plans were to update to the latest dev and merge it into upstream, so I'm more than willing to help with the development of the enocean component!

UnciaPrima commented:
It would be a really nice addition to Home Assistant if we had access to A5-11-03 as well as D2-05-00!! I've been waiting for this for so long!

Thank you very much for what you'll do!

pmorange commented:
Hi,
As I said before, I had several devices working in a custom enOcean module on my HA.
It worked with F6-02-02 (Nodon Soft Remote) and with D5-00-01 (Nodon door sensor).
If you wish I can give you the code to help with the refactor... Just tell me.

Regards.

pmorange commented:
Hi. just so you know, I opened the PR [42467](https://github.com/home-assistant/core/pull/42467) to add F6 and D5 types.

bakkerv commented:
@pmorange Nice to see this addition to HA. 

I just started to discuss with @jduquennoy how to proceed with this component. His work on the discovery would make life a lot easier, and adding more profile is always good. The thing I’m struggling with, is how to create a nice UI for it. Ideally, for each platform you could define the receiving & sending EEP. Furthermore, some teach-in and out support is needed.
Perhaps writing an extension for the front-end is needed, but we are still new to HA development, thus this will take some time ;)

DamienBraillard commented:
Hi @pmorange and @jduquennoy.

Seems we all are planning to rewrite/refactor that integration.
I had started a discussion with @MartinHjelmare on Discord when he pointed me to this PR.

I have started a new Enocean library on GitHub to better manage the protocol layer. The one currently used by HomeAssistant is not stable at all and does not use asyncio.

Next steps would have been to rewrite the Enocean Integration.

@pmorange and @jduquennoy: Can we get in touch on Discord or through another channel ? I think it would be best to bring our efforts together on this task.

bakkerv commented:
I’m unfamiliar with Discord, but I would like to join the discussion as well :)

DamienBraillard commented:
> I’m unfamiliar with Discord, but I would like to join the discussion as well :)

I'm also quite new to it.
It's what they user as communication channel for HomeAssistant. It's basically a chat. To join, go to [the Home Assistant help page](https://www.home-assistant.io/help/) and click the "Discord Chat Server" link under "Communication channels". You'll get invited into the Home Assistant group/server.

You can find me under the user "DamienBraillard#7727" on Discord. I just changed it to match my GitHub user to be found more easily (was MRD before).

pmorange commented:
@DamienBraillard I added you on Discord, do not be afraid, my nickname is different there ^^

arno0392 commented:
> You can find me under the user "DamienBraillard#7727" on Discord. I just changed it to match my GitHub user to be found more easily (was MRD before).

Hi guys,
Interested to help Enocean integration improvements, just tried to add you too (Arno).

knuedd commented:
Hi, I'd also like to help with this one. Got me a Discord account "knue" few minutes ago. What is the right branch to look at? Is there already a "Refactor Enocean part 2" underway?

pmorange commented:
> Hi, I'd also like to help with this one. Got me a Discord account "knue" few minutes ago. What is the right branch to look at? Is there already a "Refactor Enocean part 2" underway?

HI, sorry for the delay, you can use the following link to join the discord : https://discord.gg/zaGDRMBCF7

DosMaster commented:
Hi, I have through some years developed an own forked version of the Enocean component for my private use.
Mainly as i have a house full of EnOcean devices (Eltako) and to learn more about HA development.
Also included config flow to some extent, but kinda dirty...

I havn't done any work on this since March last year, I'm on 0.108 :-O
After beeing away from HA development, I now see that this component has come to live again, great news!
I will try to replace my development with #35927, #41020, #42467 and then look into if I have any code that could be of use...

Are there any other pull requests/merges that I need to be aware of?
Are there a Refactor Enocean part2 somewhere?

/Cheers, Hans

jenshenneberg commented:
@DosMaster : Have you been successful in using Home Assistant with Eltako in the current version (without your custom changes). It's quite a bit of work for me to switch over (House full of Enocean FAM14), so would be motivating to know that it works for you:-)

DosMaster commented:
Hi, sorry for the late reply, I'm loaded with work for the moment...
I havn't tried the current version with my full system, I was aiming to try this earlier this year, but havn't had the time yet...

/Hans

Will711990 commented:
Hi everybody,

I'm not a developer but I have few Enocean devices not natively supported...
Until today I have a Custom_Component adapted to support my devices, but it's time to share :)
I'm ready to help to improve the number of devices ! Could you precise me who can able to include new devices ?

Thank you !!
Will

pmorange commented:
> Hi everybody,
> 
> I'm not a developer but I have few Enocean devices not natively supported...
> Until today I have a Custom_Component adapted to support my devices, but it's time to share :)
> I'm ready to help to improve the number of devices ! Could you precise me who can able to include new devices ?
> 
> Thank you !!
> Will

Hi,
The problem is we tried that already, but they blocked any new features for this integration until there is a full refactor.
There is a Discord group if you want to know more. You can find the link in this page... But there little that has been done, I don't know if we'll be able to find courage and time one day :-d
